File: js/semio/benchmark.ts
// js/semio/benchmark.ts
import DiffForward from "../../assets/semio/diff_kit_metabolism.json";
import DiffInverse from "../../assets/semio/diff_kit_metabolism_inverted.json";
import InvalidKit from "../../assets/semio/kit_invalid.json";
import MetabolismKit from "../../assets/semio/kit_metabolism.json";
import {
    applyKitDiff,
    exportKit,
    flattenDesign,
    importKit,
    Kit,
    KitDiff,
    validateKit
} from "./semio.js";

const ITERATIONS = 3;

async function bench(name: string, fn: () => Promise<void> | void) {
    const start = performance.now();
    for (let i = 0; i < ITERATIONS; i++) {
        await fn();
    }
    const end = performance.now();
    const durationSec = (end - start) / 1000 / ITERATIONS;
    console.log(`${name},${durationSec.toFixed(6)}`);
}

function findDesign(kit: Kit, name: string, parentName?: string) {
    let parentGuid: string | undefined;
    if (parentName) {
        const p = kit.designs?.find(d => d.name === parentName);
        if (!p) throw new Error(`Parent ${parentName} not found`);
        parentGuid = p.guid;
    }
    const d = kit.designs?.find(d => d.name === name && (parentGuid ? d.parent?.guid === parentGuid : !d.parent));
    if (!d) throw new Error(`Design ${name} not found`);
    return d;
}

const kitMetabolism = MetabolismKit as unknown as Kit;
const kitInvalid = InvalidKit as unknown as Kit;
const diffForward = DiffForward as unknown as KitDiff;
const diffInverse = DiffInverse as unknown as KitDiff;

async function main() {
    // 1. Roundtrip/Metabolism
    // Pre-create a fresh zip from the JSON kit (outside the benchmark loop)
    const sourceBlob = await exportKit(kitMetabolism, new Map());
    const sourceBuffer = Buffer.from(await sourceBlob.arrayBuffer());

    await bench("Roundtrip/Metabolism", async () => {
        // Zip -> Memory -> Zip roundtrip
        const { kit, files } = await importKit(sourceBuffer);
        await exportKit(kit, files);
    });

    // 2. Diff/Metabolism
    await bench("Diff/Metabolism", () => {
        const k2 = applyKitDiff(kitMetabolism, diffForward);
        applyKitDiff(k2, diffInverse);
    });

    // 3. Flatten Design/Nakagin Capsule Tower
    const d1 = findDesign(kitMetabolism, "Nakagin Capsule Tower");
    await bench("Flatten Design/Nakagin Capsule Tower", () => {
        flattenDesign(kitMetabolism, d1.guid);
    });

    // 4. Flatten Design/Nakagin Capsule Tower/Slanted
    const d2 = findDesign(kitMetabolism, "Slanted", "Nakagin Capsule Tower");
    await bench("Flatten Design/Nakagin Capsule Tower/Slanted", () => {
        flattenDesign(kitMetabolism, d2.guid);
    });

    // 5. Flatten Design/Nakagin Capsule Tower/Twisted
    const d3 = findDesign(kitMetabolism, "Twisted", "Nakagin Capsule Tower");
    await bench("Flatten Design/Nakagin Capsule Tower/Twisted", () => {
        flattenDesign(kitMetabolism, d3.guid);
    });

    // 6. Flatten Design/Nakagin Capsule Tower/Dancing
    const d4 = findDesign(kitMetabolism, "Dancing", "Nakagin Capsule Tower");
    await bench("Flatten Design/Nakagin Capsule Tower/Dancing", () => {
        flattenDesign(kitMetabolism, d4.guid);
    });

    // 7. Flatten Design/Capsule Dream
    const d5 = findDesign(kitMetabolism, "Capsule Dream");
    await bench("Flatten Design/Capsule Dream", () => {
        flattenDesign(kitMetabolism, d5.guid);
    });

    // 8. Validation/Invalid Kit
    await bench("Validation/Invalid Kit", () => {
        validateKit(kitInvalid);
    });

    // 9. Validation/Metabolism
    await bench("Validation/Metabolism", () => {
        validateKit(kitMetabolism);
    });
}

main().catch(console.error);

File: py/semio/benchmark.py
import time
import json
import os
from semio import Kit, validateKit, flattenDesignDict, _applyDesignDiff, Type, Design

ASSETS_DIR = "../../assets/semio"
ITERATIONS = 100

def load_kit(filename: str) -> dict:
    path = os.path.join(ASSETS_DIR, filename)
    if not os.path.exists(path):
        raise FileNotFoundError(f"Asset not found: {path}")
    with open(path, "r", encoding="utf-8") as f:
        data = json.load(f)
        if "guid" in data and "uri" not in data:
            data["uri"] = data["guid"]
        for key in ["types", "designs", "files", "folders", "authors", "concepts", "models", "connectors", "pieces", "connections", "layers", "groups", "stats", "ports", "qualities", "attributes"]:
             if key not in data or data[key] is None:
                 data[key] = []
        
        # Cleanup references to IDs
        for collection in ["types", "designs", "folders"]:
            if collection in data:
                for item in data[collection]:
                    if "parent" in item and isinstance(item["parent"], dict) and "guid" in item["parent"]:
                        item["parent"] = item["parent"]["guid"]
                    if "folder" in item and isinstance(item["folder"], dict) and "guid" in item["folder"]:
                        item["folder"] = item["folder"]["guid"]

        if "types" in data:
            for t in data["types"]:
                if "models" in t:
                    for m in t["models"]:
                        if "file" in m and isinstance(m["file"], dict) and "guid" in m["file"]:
                            m["file"] = m["file"]["guid"]

        return data

def bench(name: str, func):
    start = time.perf_counter()
    for _ in range(ITERATIONS):
        func()
    end = time.perf_counter()
    duration = (end - start) / ITERATIONS
    print(f"{name},{duration:.6f}")

def find_design(kit: dict, name: str, parent_name: str = None) -> dict:
    parent_guid = None
    if parent_name:
        for d in kit.get("designs", []):
            if d.get("name") == parent_name:
                parent_guid = d.get("guid")
                break
        if not parent_guid:
            raise ValueError(f"Parent {parent_name} not found")
            
    for d in kit.get("designs", []):
        if d.get("name") == name:
            p = d.get("parent")
            if parent_guid:
                if p and p.get("guid") == parent_guid:
                    return d
            else:
                if not p:
                    return d
    raise ValueError(f"Design {name} not found")

def main():
    kit_metabolism = load_kit("kit_metabolism.json")
    kit_invalid = load_kit("kit_invalid.json")
    
    kit_obj = Kit.model_validate(kit_metabolism)
    if kit_obj.types is None and "types" in kit_metabolism:
        kit_obj.types = [Type.model_validate(t) for t in kit_metabolism["types"]]
    
    kit_invalid_obj = Kit.model_validate(kit_invalid)
    if kit_invalid_obj.types is None and "types" in kit_invalid:
        kit_invalid_obj.types = [Type.model_validate(t) for t in kit_invalid["types"]]
    
    # 1. Roundtrip/Metabolism
    def test_roundtrip():
        s = kit_obj.model_dump_json()
        Kit.model_validate_json(s)
        
    bench("Roundtrip/Metabolism", test_roundtrip)
    
    # 2. Flatten Design/Nakagin Capsule Tower
    d1 = find_design(kit_metabolism, "Nakagin Capsule Tower")
    def test_flatten_nakagin():
        diff = flattenDesignDict(kit_metabolism, d1["guid"])
        _applyDesignDiff(d1, diff)
        
    bench("Flatten Design/Nakagin Capsule Tower", test_flatten_nakagin)
    
    # 3. Flatten Design/Nakagin Capsule Tower/Slanted
    d2 = find_design(kit_metabolism, "Slanted", "Nakagin Capsule Tower")
    def test_flatten_nakagin_slanted():
        diff = flattenDesignDict(kit_metabolism, d2["guid"])
        _applyDesignDiff(d2, diff)
        
    bench("Flatten Design/Nakagin Capsule Tower/Slanted", test_flatten_nakagin_slanted)

    # 4. Flatten Design/Nakagin Capsule Tower/Twisted
    d3 = find_design(kit_metabolism, "Twisted", "Nakagin Capsule Tower")
    def test_flatten_nakagin_twisted():
        diff = flattenDesignDict(kit_metabolism, d3["guid"])
        _applyDesignDiff(d3, diff)
        
    bench("Flatten Design/Nakagin Capsule Tower/Twisted", test_flatten_nakagin_twisted)

    # 5. Flatten Design/Nakagin Capsule Tower/Dancing
    d4 = find_design(kit_metabolism, "Dancing", "Nakagin Capsule Tower")
    def test_flatten_nakagin_dancing():
        diff = flattenDesignDict(kit_metabolism, d4["guid"])
        _applyDesignDiff(d4, diff)
        
    bench("Flatten Design/Nakagin Capsule Tower/Dancing", test_flatten_nakagin_dancing)

    # 6. Flatten Design/Capsule Dream
    d5 = find_design(kit_metabolism, "Capsule Dream")
    def test_flatten_capsule_dream():
        diff = flattenDesignDict(kit_metabolism, d5["guid"])
        _applyDesignDiff(d5, diff)
        
    bench("Flatten Design/Capsule Dream", test_flatten_capsule_dream)
    
    # 7. Validation/Invalid Kit
    def test_validate_invalid():
        validateKit(kit_invalid_obj)
        
    bench("Validation/Invalid Kit", test_validate_invalid)
    
    # 8. Validation/Metabolism
    def test_validate_metabolism():
        validateKit(kit_obj)
        
    bench("Validation/Metabolism", test_validate_metabolism)

if __name__ == "__main__":
    main()
